daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SequenceGroup.scala

```diff
@@ -164,9 +165,10 @@ abstract class SequenceGroupTermBase(xml: Node, lexicalParent: SchemaComponent,
   protected final lazy val checkIfNonEmptyAndDiscrimsOrAsserts: Unit = {
     val msg = "Counterintuitive placement detected. Wrap the discriminator or assert " +
       "in an empty sequence to evaluate before the contents."
-    if (groupMembers.nonEmpty && discriminatorStatements.nonEmpty)
+    // Only show warning if the testKind=expression, see DFDL spec Section 9.5 Evaluation Order for Statement Annotations
+    if (groupMembers.nonEmpty && discriminatorStatements.filter(_.testKind == TestKind.Expression).nonEmpty)
```

Comment:
Might be a little more clear to do `discriminatorStatements.exists(_.testKind == TestKind.Expression)`

Reply:
Agreed.  I always forget about exists()